### PR TITLE
Sync progress enhancement work.

### LINKED ISSUE #3294: Updating Sync button on the Navigation menu based on Sync Status
**Describe the feature request.**
We would like to update the UI/UX for manual sync. This would be done by updating the UI to allow users to know when they have unsynced data, and track sync progress and final status.

**Additional context**
- [ ]  Update the sync button as shown below:

**Acceptance criteria**

- [ ] When there are unsynced data, the sync button on the side navigation has a warning icon and changes colour to yellow as shown in the image (b).
- [ ] When we do not have unsynced data, (all resources have successfully been synced to the server), the sync button should appear as shown in the image (a).
- [ ] The time last successful sync was  conducted is also shown.

**Area path**
1. Login to the app
2. Open the Navigation bar
3. If there is no unsynced data, sync button should appear as on image (a)
4. If there are unsynced data, the sync button should appear as on image (b)

**Implementation plan (For Engineers)**
The plan for implementing the solution e.g. via a _description_ or a _checklist_ for the various ordered tasks that will need to be completed.
_i.e. Describe how you intend to solve the problem_

### LINKED ISSUE #3295: Add a Sync Progress component for subsequent sync
**Problem we are trying to solve**
Currently, when a user is conducting sync, initial or subsequent, they are hard pressed to know the progress of the sync until it is completed. This does not make for a good user experience; the user may abandon the sync process or sometimes think there is somethign wrong with the application; to them, it seems like there is nothing happening. 

**User Story**
As a user, I want to see the sync progress when interacting with the app. I want to decide whether to proceed with sync as I do other things or simply wait for it to complete. I want to be aware whether there are processes happening on the app even when it seems like nothing is happening. This is especially necessary when I have a lot of data on my device that may take some time to complete syncing. 

**Describe the feature request.**
We would like to update the UI/UX for manual sync. This would be done by updating the UI to allow users to know when they have unsynced data, and track sync progress and final status. 

This ticket is to specifically add tracking and updating the subsequent sync status. 

**Implementation**
- [ ]  When the sync button is tapped, image a (see screenshot below)
  - [ ] User is navigated to the default register with a persistent sync progress component as shown in image (b)
  - [ ] Sync progress component should also be shown on the nav menu as seen in image (d) below. The sync component replaces the manual sync button when sync is in progress.
  
- [x] Sync progress component should:
  - [x] Indicate the percentage of sync
  - [x] Have a progress bar that reflects the % the sync is at. 
  - [x] Indicate time remaining for sync to complete.
  - [x] Have the spinner icon spinning so the user doesn’t assume the app/sync froze.
  - [x] Cancel button to cancel the sync at any time as shown.
  - [x] Minimize component button as shown above.
  - [x] After 10-seconds since starting sync, the sync progress component will be automatically minimized as shown on the image.
- [ ] When the sync component is minimized show:
   - [x] A progress bar
  - [x] A maximize button
  - [ ] A spinner icon
  
- [ ] When the cancel button is tapped, display a confirmation dialog:
  - [ ] keep syncing dismisses the dialog
  - [ ] Stop sync cancels sync
  
  

- [ ] When sync is complete (All data has been synced to the server):
  - [ ] Sync complete component is displayed for 10 seconds after sync is complete.
  - [ ] The sync complete component will be sustained until there is new unsynced data in the app. 
  - [ ] If there is unsynced data, this component is replaced by the manual sync button. 

image

- [ ] Not all data has been synced to the server.
    - [ ] Error message is displayed.
    - [ ] The sync component has a retry button and tapping the retry should restart the sync.

**Acceptance criteria**
All the items above work as they should.

**Area path**
1. Login to the app
2. Open Navigation bar
3. If there is no unsyced data, sync button should appear as on image (a)
4. If there are unsyced data, sync button should appear as on image (b)

**Implementation plan (For Engineers)**
The plan for implementing the solution e.g. via a _description_ or a _check list_ for the various ordered tasks that will need to be completed.
_i.e. Describe how you intend to solve the problem_

### PR DESCRIPTION
**Pull request by Lentumunai-Mark**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #3295
Fixes #3294

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [ ] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by ellykits** [comment]
@Lentumunai-Mark Can you share screenshot or animated `gif` for this?

**Comment by ellykits** [comment]
@Lentumunai-Mark on the UI I still get the snackbar notifications? Was this intended?

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterScreen.kt [hunk 1]
```diff
@@ -67,35 +104,80 @@ const val REGISTER_CARD_TEST_TAG = "registerCardListTestTag"
...
+    registerUiState.isFirstTimeSync,
+  ) {
+    when {
+      appUiState?.isSyncCompleted == SyncStatus.SUCCEEDED &&
+        appUiState.progressPercentage == 100 -> {
+        backgroundColor = Color(0xFF1DB11B)
+        showSyncBar = true
+        applyBackgroundColor = true
+        coroutineScope.launch {
+          delay(60000L)
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterScreen.kt:145]: Any reason for adding this delay?

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/main/AppMainUiState.kt [hunk 2]
```diff
@@ -47,5 +53,15 @@ fun appMainUiStateOf(
     languages = languages,
     navigationConfiguration = navigationConfiguration,
     registerCountMap = registerCountMap,
+    progressPercentage = progressPercentage,
+    isSyncUpload = isSyncUpload,
+    isSyncCompleted = syncStatus,
   )
 }
+
+enum class SyncStatus {
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/main/AppMainUiState.kt:62]: Would it be possible to use `CurrentSyncJobStatus` directly here? Do you have any explanation as to why this enum is required?

**Review by ellykits — CHANGES_REQUESTED** [review]

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterScreen.kt (hunk 1 above)
Comment (Lentumunai-Mark) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterScreen.kt:145]: @ellykits 
We need the delay after the sync operation is completed (whether it succeeds or fails) to give users enough time to notice the sync result before the UI reverts to its default state

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/main/AppMainUiState.kt (hunk 2 above)
Comment (Lentumunai-Mark) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/main/AppMainUiState.kt:62]: @ellykits 
thanks for the suggestion
looking into this.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterScreen.kt (hunk 1 above)
Comment (ellykits) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/register/RegisterScreen.kt:145]: What happens when the delay is not define?

**Comment by ellykits** [comment]
Fix design inconsistencies @Lentumunai-Mark

**Comment by ellykits** [comment]
@Lentumunai-Mark  Observe the sync states for geowidget map that also uses the app drawer

**Review by ellykits — APPROVED** [review]

**Pull request merged**